test/compiler/codegen.jl

```diff
@@ -581,13 +581,13 @@ end
     end
     @test occursin("llvm.julia.gc_preserve_begin", get_llvm(f3, Tuple{Bool}, true, false, false))
 
-    # unions of immutables (JuliaLang/julia#39501)
+    # PhiNode of unions of immutables (JuliaLang/julia#39501)
     function f2(cond)
         val = cond ? 1 : 1f0
         GC.@preserve val begin end
         return cond
     end
-    @test !occursin("llvm.julia.gc_preserve_begin", get_llvm(f2, Tuple{Bool}, true, false, false))
+    @test occursin("llvm.julia.gc_preserve_begin", get_llvm(f2, Tuple{Bool}, true, false, false))
```

Comment:
What exactly are we preserving here? There shouldn't be any legal way to get a reference to this box.

Reply:
Prior to optimization, there are a pair of phi node here (one for unboxed unions, one for Vboxed objects). We don't know which one is being used later by the user, for the sake of this emission step. For example, given:

```
julia> function f2(cond)
           val = cond ? 1 : []
           GC.@preserve val begin; GC.gc(); end
           return cond
       end
f2 (generic function with 1 method)

julia> code_llvm(f2, (Bool,), optimize=true)
```

There was no root for `[]` before this change, because codegen was considering it to be impossible to reach that at runtime.

Reply:
Can we change the test to have one mutable value then?

Reply:
The same result applies for immutable though, you might just need a different form to allocate it. But there is otherwise no difference for mutables

Reply:
`@GC.preserve` has no effect on an immutable value, because the lifetime is not semantically observable, so it would be legal to remove it. It's not a good test if the opposite is also a valid result.

Reply:
The lifetime of the objects inside are observable though for immutable. We don't consistently distinguish here between the isbits case and the others though, which I think you make a strong argument for being a legal optimization to do if we want to followup with that. It doesn't seem particularly important to me though as a case.

Reply:
I don't think the optimization is important per-se, I would just like the test to be something that still works if the optimization is implemented.